### ISSUE TITLE
Issue when parsing ISO8601 date string on GAE

### ISSUE THREAD

**Issue by aullrich2013** [dateutil/dateutil#64]
I'm using version 2.4.1 in both a local GAE environment as well as in a local ipython instance. 
In both I attempt to parse the string using:

```
asdf = dateutil.parser.parse(u'2014-06-01T00:00:00.000Z')
print asdf.tzinfo
```

On the local machine, this properly shows tzutc as the result
When using the GAE environment, this instead returns a datetime with tzlocal

It looks like this is being set in parser.py at line 424 (425 after my pdb insertion)

```
(Pdb) l
420                         raise ValueError("Offset must be tzinfo subclass, "
421                                          "tz string, or int offset.")
422                     ret = ret.replace(tzinfo=tzinfo)
423                 elif res.tzname and res.tzname in time.tzname:
424                     import pdb; pdb.set_trace()
425  ->                 ret = ret.replace(tzinfo=tz.tzlocal())
```

I'm not sure what the issue is but this seems very strange.

**Comment by pganssle** [comment]
@aullrich2013 Is the GAE environment using UTC as its local time? Because `time.tzname` retrieves the names of the time zones of the local time. Aside from ambiguous times during DST switchovers (see #57), it seems like there shouldn't be a difference between using the named time zone and `tzlocal()`, though I'm not entirely sure what the advantage to using `tzlocal()` is - perhaps there are more common conditions where it is more stable.

**Comment by aullrich2013** [comment]
Good question. GAE does use UTC as its local time. time.tzname gives ('UTC', 'UTC'). 

I've got it now. Thanks!

**Comment by pganssle** [comment]
OK, closing as not a bug.

**Comment by nbonnotte** [comment]
I'm encountering the same issue, and for me it really looks like a bug.

I'm writing a piece of code that will run on different environments and I have no control on the time zone. Sometimes it's UTC, sometimes it is not.

I need to check that a given string represent a datetime in the UTC timezone.

On my local machine (Europe/Paris), it's easy:
```
In [2]: tzlocal.get_localzone()
Out[2]: <DstTzInfo 'Europe/Paris' LMT+0:09:00 STD>

In [4]: parser.parse(u'2016-08-24T12:14:39.783776+0:00')
Out[4]: datetime.datetime(2016, 8, 24, 12, 14, 39, 783776, tzinfo=tzutc())

In [5]: parser.parse(u'2016-08-24T12:14:39.783776+0:00').tzinfo == tz.tzutc()
Out[5]: True
```

But when the environment is set to UTC, it fails:
```
In [2]: tzlocal.get_localzone()
Out[2]: <UTC>

In [4]: parser.parse(u'2016-08-24T12:14:39.783776+0:00')
Out[4]: datetime.datetime(2016, 8, 24, 12, 14, 39, 783776, tzinfo=tzlocal())

In [7]: parser.parse(u'2016-08-24T12:14:39.783776+0:00').tzinfo == tz.tzutc()
Out[7]: False
```

The only solution I have found is this convoluted check:
```
In [3]: dt = parser.parse(u'2016-08-24T12:14:39.783776+0:00')

In [4]: dt.tzinfo == tz.tzutc() or (dt.tzinfo == tz.tzlocal() and tzlocal.get_localzone() == pytz.utc)
Out[4]: True
```

**Comment by pganssle** [comment]
@nbonnotte OK, I'm going to reopen this to look into it, but I think both modes are *defensible*, and when both are defensible I think "don't change anything" tends to rule out so that we're not changing behavior that people may be relying on.

I'm not sure why you want to check the time zone of the parsed date, though. If you just want the time zone in UTC, you don't actually need to check if it's already in UTC, you can just do this:

```
from dateutil import tz
dt_in_utc = dt.astimezone(tz.utc())
```

Then there's no ambiguity - it's in UTC. If you have some control structure that depends on whether or not this datetime was originally represented in UTC, I don't know that there's a non-convoluted way to do it, though there are many simple ways to do it depending on what edge cases you need to account for. You should be able to just check that `dt.tzinfo.utcoffset() == 0`, which would work unless you care about differentiating between UTC and something like WET, which is UTC+0 sometimes, but transitions to WEST, UTC+1 during the summer. This would only come up if your application runs on computers representing timestamps like `2015-06-01 12:04:33 WET` and their local time zone is set to `WET` or `WEST`.

In any case, I think it may be a separate issue, but I will look into the possibility of modifying `tz.tzlocal` such that if it is effectively UTC that it compares equal to `tz.tzutc`.

**Comment by pganssle** [comment]
The "UTC gets parsed as local part" main issue is #349 now.

**Comment by pganssle** [comment]
The equality issue is closed in #598.
